File: aiplaylist_common_files/src/main/java/net/ssingh/aiplaylist_common_files/model/dto/request/playlist/GeneratePlaylistRequest.java
package net.ssingh.spotifyservice.model.dto.request.playlist;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import net.ssingh.spotifyservice.model.entity.generic.Artist;
import net.ssingh.spotifyservice.model.entity.generic.Track;

import java.io.Serializable;
import java.util.List;

@Data
@AllArgsConstructor
@NoArgsConstructor
public class GeneratePlaylistRequest implements Serializable {
    @JsonProperty(required = true)
    String userPrompt;
    @JsonProperty(required = true)
    int playlistLength;

    @JsonProperty(required = false)
    List<Track<Artist>> tracks;
    @JsonProperty(required = false)
    List<Artist> artists;

    @JsonProperty(required = false)
    Double temperature;

    @JsonProperty(required = false)
    Double nicheSlider;
}

File: next-frontend/src/models/playlist/GeneratePlaylistRequest.ts
import {Track} from "@/models/track/Track";
import {Artist} from "@/models/artist/Artist";
import {Playlist} from "@/models/playlist/Playlist";

export default interface GeneratePlaylistRequest {
    query: string;
    playlistLength: number;
    temperature: number;
    nicheSlider: number;
    playlist?: Playlist<Track<Artist>> | null;
    selectedTracks?: Track<Artist>[] | null;
    selectedArtists?: Artist[] | null;
}

export const GeneratePlaylistBodyString = (req: GeneratePlaylistRequest): string => JSON.stringify({
    userPrompt: req.query,
    playlistLength: req.playlistLength,
    temperature: req.temperature,
    nicheSlider: req.nicheSlider,
    tracks: [...((req.playlist?.tracks as Track<Artist>[]) || []), ...req.selectedTracks!],
    artists: req.selectedArtists as Artist[],
});